### PR TITLE
Pass arguments to userspace programs.

### PR DESCRIPTION
**Pull request by pj1031999**
Add option to pass arguments to init program using launch. Everything after --
will be passed as arguments.
Example command:
./launch -d init=/bin/ls -- / /usr

### CODE REVIEW

File: sys/kern/main.c [hunk 1]
```diff
@@ -17,7 +19,20 @@ int kmain(void) {
   proc_add(proc_create(thread_self(), NULL));
 
   if (init) {
-    run_program(init, (char *[]){init, NULL}, (char *[]){NULL});
+    int ninit_args;
+    char **init_args = kenv_get_init_args(&ninit_args);
+
+    if (init_args) {
+      char **args = kmalloc(M_TEMP, sizeof(char *) * (ninit_args + 2), 0);
```
Comment (xThaid) [inline sys/kern/main.c]: Shouldn't this memory leak be handled?

File: sys/mips/malta.c [hunk 2]
```diff
@@ -122,6 +125,19 @@ char *kenv_get(const char *key) {
...
+    char *arg = _kenv.argv[i];
+    if ((strncmp("--", arg, 2) == 0)) {
+      if (n) {
+        *n = _kenv.argc - i - 1;
+      }
+      return _kenv.argv + i + 1;
+    }
+  }
+  return NULL;
+}
```
Comment (jpszczolowski) [inline sys/mips/malta.c]: An idea: changing `return NULL;` to `*n = 0; return NULL;` would enable you to merge two cases in file `sys/kern/main.c`: `if (init_args) { ... } else { ... }` into just one. Then you could delete:
 ```
run_program(init, (char *[]){init, NULL}, (char *[]){NULL});
```
and always call:
```
run_program(init, args, (char *[]){NULL});
```

File: sys/kern/main.c [hunk 3]
```diff
@@ -5,9 +5,11 @@
 #include <sys/proc.h>
 #include <sys/thread.h>
 #include <sys/ktest.h>
+#include <sys/malloc.h>
 
 /* Borrowed from mips/malta.c */
 char *kenv_get(const char *key);
+char **kenv_get_init_args(int *n);
```
Comment (jpszczolowski) [inline sys/kern/main.c]: I find it a little bit confusing that `kenv_get_init_args()` returns the result both as a regular return value and through a pointer. But maybe it's just nitpicking.

File: sys/kern/main.c (hunk 1 above)
Comment (pj1031999) [reply sys/kern/main.c]: Done. (d6fd9d470780afc6d27bc4f28035418f04d7a828)

File: sys/mips/malta.c (hunk 2 above)
Comment (pj1031999) [reply sys/mips/malta.c]: Done. (d6fd9d470780afc6d27bc4f28035418f04d7a828)

File: sys/kern/main.c (hunk 3 above)
Comment (pj1031999) [reply sys/kern/main.c]: From my point of view it is the simplest solution in this case. But if you know something more elegant (and not more complicated) feel free to share your idea.

**Pull request merged**